### PR TITLE
SecurityGroups creation and removal should be associated to TierInstance creation and removal

### PR DESCRIPTION
**Pull request by jmmovilla**
@jframos @hmunfru

### CODE REVIEW

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/TierInstanceManager.java [hunk 1]
```diff
@@ -124,7 +125,8 @@ void create(ClaudiaData claudiaData, TierInstance tierInstance, EnvironmentInsta
      * @throws InvalidEntityException
      * @throws InfrastructureException 
      */
-    TierInstance create(ClaudiaData claudiaa, String envName, TierInstance tierInstance) throws InvalidEntityException, InfrastructureException;
+    TierInstance create(ClaudiaData claudiaa, String envName, TierInstance tierInstance) throws InvalidEntityException,
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/TierInstanceManager.java]: 121 characters, i see here.

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/TierManager.java [hunk 2]
```diff
@@ -137,5 +138,13 @@ void addSecurityGroupToProductRelease(ClaudiaData claudiaData, Tier tier, Produc
      * @throws InvalidEntityException
      */
     void updateTier(ClaudiaData data, Tier tierold, Tier tiernew) throws InvalidEntityException, EntityNotFoundException, AlreadyExistsEntityException;
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/TierManager.java]: more than 120 chars.

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 3]
```diff
@@ -409,7 +417,8 @@ public void destroy(ClaudiaData claudiaData, EnvironmentInstance envInstance) th
 
     }
 
-    private void deletePaaSDB(ClaudiaData claudiaData, EnvironmentInstance envInstance, TierInstance tierInstance, boolean error) throws InvalidEntityException {
+    private void deletePaaSDB(ClaudiaData claudiaData, EnvironmentInstance envInstance, TierInstance tierInstance, boolean error)
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: more than 120 chars.

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 4]
```diff
@@ -417,10 +426,29 @@ private void deletePaaSDB(ClaudiaData claudiaData, EnvironmentInstance envInstan
         try {
             infrastructureManager.deleteNetworksInTierInstance(claudiaData, tierInstance);
         } catch (Exception e) {
-            log.error(e.getMessage());
+        	log.error(e.getMessage());
             error = true;
+        } finally {
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: the coding style of this block code should be corrected.

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 5]
```diff
@@ -442,48 +472,12 @@ public void remove(TierInstance tierInstance) throws InvalidEntityException {
...
-    public void setTierInstanceDao(TierInstanceDao tierInstanceDao) {
-        this.tierInstanceDao = tierInstanceDao;
-    }
-
-    public void setTierManager(TierManager tierManager) {
-        this.tierManager = tierManager;
-    }
-
-    public TierInstance update(ClaudiaData claudiaData, String envName, TierInstance tierInstance)
+        public TierInstance update(ClaudiaData claudiaData, String envName, TierInstance tierInstance)
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: the indentation of the method is not correct.

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 6]
```diff
@@ -522,5 +516,180 @@ public void update(ClaudiaData claudiaData, TierInstance tierInstance, Environme
...
+     * It creates the specified security groups.
+     * 
+     * @param claudiaData
+     * @param tierInstance
+     * @return
+     * @throws InvalidSecurityGroupRequestException
+     * @throws EntityNotFoundException
+     */
+
+    private TierInstance createSecurityGroups(ClaudiaData claudiaData, TierInstance tierInstance) throws InvalidSecurityGroupRequestException,
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: >120 chars.

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 7]
```diff
@@ -522,5 +516,180 @@ public void update(ClaudiaData claudiaData, TierInstance tierInstance, Environme
...
+            EntityNotFoundException {
+        if ((systemPropertiesProvider.getProperty(SystemPropertiesProvider.CLOUD_SYSTEM).equals("FIWARE")
+                && claudiaData.getVdc() != null && claudiaData.getVdc().length() > 0)) {
+
+            SecurityGroup securityGroup = generateSecurityGroup(claudiaData, tierInstance);
+            try {
+                securityGroup = securityGroupManager.create(tierInstance.getTier().getRegion(), claudiaData.getUser().getToken(),
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: >120 chars.

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 8]
```diff
@@ -522,5 +516,180 @@ public void update(ClaudiaData claudiaData, TierInstance tierInstance, Environme
...
+                        + securityGroup.getName() + " " + e.getMessage(), e);
+
+            }
+            tierInstance.setSecurityGroup(securityGroup);
+        }
+        return tierInstance;
+
+    }
+    
+   private SecurityGroup generateSecurityGroup(ClaudiaData claudiaData, TierInstance tierInstance) throws EntityNotFoundException {
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: > 120 chars.

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 9]
```diff
@@ -522,5 +516,180 @@ public void update(ClaudiaData claudiaData, TierInstance tierInstance, Environme
...
+
+        SecurityGroup securityGroup = new SecurityGroup();
+        securityGroup.setName("sg_" + claudiaData.getService() + "_" + claudiaData.getVdc() + "_" + tierInstance.getName());
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: gt 120 chars.

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 10]
```diff
@@ -522,5 +516,180 @@ public void update(ClaudiaData claudiaData, TierInstance tierInstance, Environme
...
+
+            for (ProductRelease productRelease : tierInstance.getTier().getProductReleases()) {
+                getRulesFromProduct(productRelease, rules);
+            }
+        }
+        securityGroup.setRules(rules);
+        return securityGroup;
+    }
+    
+   	   	private List<Rule> getDefaultRules() {
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: correct the indentation of the method name.

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 11]
```diff
@@ -67,14 +73,16 @@
     private ProductInstanceManager productInstanceManager;
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: please provide javadoc for the different methods that you have in the class.

File: rest-api/src/main/java/com/telefonica/euro_iaas/paasmanager/rest/validation/TierResourceValidatorImpl.java [hunk 12]
```diff
@@ -167,7 +167,7 @@ private void validataDefaultTier(TierDto tierDto) throws InvalidEntityException
 
         if (!(tierDto.getMinimumNumberInstances() <= tierDto.getInitialNumberInstances() && tierDto
                 .getMaximumNumberInstances() >= tierDto.getInitialNumberInstances())) {
-            String men = "Error in the Number initial " + tierDto.getInitialNumberInstances() + " with number min "
+            String men = "The number of replicas is not valid. Error in the Number initial " + tierDto.getInitialNumberInstances() + " with number min "
```
Comment (flopezag) [inline rest-api/src/main/java/com/telefonica/euro_iaas/paasmanager/rest/validation/TierResourceValidatorImpl.java]: gt 120 chars.

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/SecurityGroupDao.java [hunk 13]
```diff
@@ -35,6 +36,8 @@
  */
 public interface SecurityGroupDao extends BaseDAO<SecurityGroup, String> {
 
-    SecurityGroup updateSecurityGroupId(String securityGroupId, SecurityGroup securityGroup)
+	SecurityGroup loadWithRules(String name) throws EntityNotFoundException;
+
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/SecurityGroupDao.java]: please, check the indentation of this lines...

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/impl/TierDaoJpaImpl.java [hunk 14]
```diff
@@ -228,11 +231,29 @@ public Tier loadComplete(Tier newTier) throws EntityNotFoundException {
             tier = (Tier) query.getResultList().get(0);
             tier.getNetworks();
             tier.getProductReleases();
-            tier.getSecurityGroup();
+            //tier.getSecurityGroup();
         } catch (Exception e) {
             String message = " Tier don't exist in database ";
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/impl/TierDaoJpaImpl.java]: Tier does not exist in database

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 15]
```diff
@@ -417,10 +426,29 @@ private void deletePaaSDB(ClaudiaData claudiaData, EnvironmentInstance envInstan
         try {
             infrastructureManager.deleteNetworksInTierInstance(claudiaData, tierInstance);
         } catch (Exception e) {
-            log.error(e.getMessage());
+        	log.error(e.getMessage());
             error = true;
+        } finally {
+        	//Deleting SG
+            log.info("Deleting security group from in tier " + tierInstance.getName() + " in TierInstance");
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: Assume that you want to say: Deleting security group from tier...

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 16]
```diff
@@ -417,10 +426,29 @@ private void deletePaaSDB(ClaudiaData claudiaData, EnvironmentInstance envInstan
...
             infrastructureManager.deleteNetworksInTierInstance(claudiaData, tierInstance);
         } catch (Exception e) {
-            log.error(e.getMessage());
+        	log.error(e.getMessage());
             error = true;
+        } finally {
+        	//Deleting SG
+            log.info("Deleting security group from in tier " + tierInstance.getName() + " in TierInstance");
+            if (tierInstance.getSecurityGroup() != null && !tierInstance.getVdc().isEmpty()) {
+            	SecurityGroup securityGroup=null;
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: SecurityGroup securityGroup = null;

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 17]
```diff
@@ -417,10 +426,29 @@ private void deletePaaSDB(ClaudiaData claudiaData, EnvironmentInstance envInstan
...
+            	try {
+        			securityGroup = securityGroupDao.loadWithRules(tierInstance.getSecurityGroup().getName());
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: The second time that you want to access to tierInstance.getSecurityGroup(), why do not call something like this:

Something securityGroup = tierInstance.getSecurityGroup();

And then make the if on line 434 and this line 437 with this variable.

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/SecurityGroupManagerImpl.java [hunk 18]
```diff
@@ -67,6 +68,23 @@ public SecurityGroup create(String region, String token, String vdc, SecurityGro
         return securityGroupDB;
     }
 
+    public SecurityGroup createInOpenstack(String region, String token, String vdc, SecurityGroup securityGroup)
+            throws InvalidEntityException, AlreadyExistsEntityException, InfrastructureException {
+
+        String idSecurityGroup = firewallingClient.deploySecurityGroup(region, token, vdc, securityGroup);
+        log.info("Create security group " + securityGroup.getName() + " with idSecurityGroup " + idSecurityGroup);
+
+        if (securityGroup.getRules() != null) {
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/SecurityGroupManagerImpl.java]: I assume that if you are here, the securityGroup is not null. Is it correct?

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/SecurityGroupManagerImpl.java [hunk 19]
```diff
@@ -67,6 +68,23 @@ public SecurityGroup create(String region, String token, String vdc, SecurityGro
...
     }
 
+    public SecurityGroup createInOpenstack(String region, String token, String vdc, SecurityGroup securityGroup)
+            throws InvalidEntityException, AlreadyExistsEntityException, InfrastructureException {
+
+        String idSecurityGroup = firewallingClient.deploySecurityGroup(region, token, vdc, securityGroup);
+        log.info("Create security group " + securityGroup.getName() + " with idSecurityGroup " + idSecurityGroup);
+
+        if (securityGroup.getRules() != null) {
+            for (Rule rule : securityGroup.getRules()) {
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/SecurityGroupManagerImpl.java]: You can optimise this operation if previously to the if, something like 

X x = securityGroup.getRules()

If ( x != null ) {
    for (Rule rule : x) {
       ...
   }
}

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/SecurityGroupManagerImpl.java [hunk 20]
```diff
@@ -112,15 +130,15 @@ public SecurityGroup insert(SecurityGroup securityGroup) throws AlreadyExistsEnt
...
-        if (securityGroup.getRules().isEmpty()) {
+    	log.info("Destroying securitygroup: " + secGroup.getName());
+        if (secGroup.getRules().isEmpty()) {
             log.warn("There is not any rule associated to the security group");
         } else {
-            List<Rule> rules = securityGroup.cloneRules();
-            securityGroup.setRules(null);
-            securityGroupDao.update(securityGroup);
+            List<Rule> rules = secGroup.cloneRules();
+        	secGroup.setRules(null);
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/SecurityGroupManagerImpl.java]: wrong indentation...

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 21]
```diff
@@ -99,10 +107,25 @@ public TierInstance create(ClaudiaData data, String envName, TierInstance tierIn
             throw new InvalidEntityException("Error to load the Tier " + tierInstance.getTier().getName() + " : "
                     + e.getMessage());
         }
-        tierInstanceDB.setTier(tierDB);
-
-        if (tierInstance.getProductInstances() != null) {
-            for (ProductInstance productInstance : tierInstance.getProductInstances()) {
+        
+        try {
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: check the indentation of this try.catch block code.

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 22]
```diff
@@ -99,10 +107,25 @@ public TierInstance create(ClaudiaData data, String envName, TierInstance tierIn
...
+        } catch (EntityNotFoundException enfe) {
+        	String men = "ProductRelease in tier " + tierInstance.getTier().getName() + "does not exist"; 
+        	log.error(men);
+            throw new InvalidEntityException(men + " : " + enfe.getMessage());
+        }
+        
+       tierInstanceDB.setTier(tierDB);
+       tierInstanceDB.setSecurityGroup(tierInstance.getSecurityGroup());
+       
+       if (tierInstance.getProductInstances() != null) {
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: You can optimise this operation in line with the previous comment. Create a auxiliar variable an only call the operation once...

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 23]
```diff
@@ -229,10 +252,17 @@ public void create(ClaudiaData claudiaData, TierInstance tierInstance, Environme
     public void delete(ClaudiaData claudiaData, TierInstance tierInstance, EnvironmentInstance envInstance)
             throws InfrastructureException, InvalidEntityException, EntityNotFoundException {
 
-        infrastructureManager.deleteVMReplica(claudiaData, tierInstance);
-        envInstance.removeTierInstance(tierInstance);
-        environmentInstanceManager.update(envInstance);
-        remove(tierInstance);
+    	log.error("before infrastructureManager.deleteVMReplica");
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: Are you sure that you need to put all those log.error messages?

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 24]
```diff
@@ -522,5 +516,180 @@ public void update(ClaudiaData claudiaData, TierInstance tierInstance, Environme
...
+        if ((systemPropertiesProvider.getProperty(SystemPropertiesProvider.CLOUD_SYSTEM).equals("FIWARE")
+                && claudiaData.getVdc() != null && claudiaData.getVdc().length() > 0)) {
+
+            SecurityGroup securityGroup = generateSecurityGroup(claudiaData, tierInstance);
+            try {
+                securityGroup = securityGroupManager.create(tierInstance.getTier().getRegion(), claudiaData.getUser().getToken(),
+                        claudiaData.getVdc(), securityGroup);
+                tierInstance.setSecurityGroup(securityGroup);
+            } catch (InvalidEntityException e) {
+                log.error("It is not posssible to create the security group " + securityGroup.getName() + " "
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: It is not possible to create the security group

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 25]
```diff
@@ -522,5 +516,180 @@ public void update(ClaudiaData claudiaData, TierInstance tierInstance, Environme
...
+                        + e.getMessage());
+                throw new InvalidSecurityGroupRequestException("It is not posssible to create the security group "
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: Idem

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 26]
```diff
@@ -522,5 +516,180 @@ public void update(ClaudiaData claudiaData, TierInstance tierInstance, Environme
...
+                        + securityGroup.getName() + " " + e.getMessage(), e);
+            } catch (InvalidEnvironmentRequestException e) {
+
+                log.error("It is not posssible to create the security group " + securityGroup.getName() + " "
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: I suggest to change the description of the error in order to facilitate the traceability of the errors message...

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 27]
```diff
@@ -522,5 +516,180 @@ public void update(ClaudiaData claudiaData, TierInstance tierInstance, Environme
...
+                        + e.getMessage());
+                throw new InvalidSecurityGroupRequestException("It is not posssible to create the security group "
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: Idem

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 28]
```diff
@@ -522,5 +516,180 @@ public void update(ClaudiaData claudiaData, TierInstance tierInstance, Environme
...
+                        + securityGroup.getName() + " " + e.getMessage(), e);
+            } catch (AlreadyExistsEntityException e) {
+                log.error("It is not posssible to create the security group " + securityGroup.getName() + " "
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: Idem

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 29]
```diff
@@ -522,5 +516,180 @@ public void update(ClaudiaData claudiaData, TierInstance tierInstance, Environme
...
+                        + e.getMessage());
+                throw new InvalidSecurityGroupRequestException("It is not posssible to create the security group "
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: Idem

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 30]
```diff
@@ -522,5 +516,180 @@ public void update(ClaudiaData claudiaData, TierInstance tierInstance, Environme
...
+                        + securityGroup.getName() + " " + e.getMessage(), e);
+            } catch (InfrastructureException e) {
+                log.error("It is not posssible to create the security group " + securityGroup.getName() + " "
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: Idem

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 31]
```diff
@@ -522,5 +516,180 @@ public void update(ClaudiaData claudiaData, TierInstance tierInstance, Environme
...
+                        + e.getMessage());
+                throw new InvalidSecurityGroupRequestException("It is not posssible to create the security group "
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: Idem

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 32]
```diff
@@ -522,5 +516,180 @@ public void update(ClaudiaData claudiaData, TierInstance tierInstance, Environme
...
+
+        SecurityGroup securityGroup = new SecurityGroup();
+        securityGroup.setName("sg_" + claudiaData.getService() + "_" + claudiaData.getVdc() + "_" + tierInstance.getName());
+
+        log.info("Generate Object Security Group " + "sg_" + claudiaData.getService() + "_" + claudiaData.getVdc() 
+        		+ "_" + tierInstance.getTier().getRegion() + "_" + tierInstance.getName());
+
+        List<Rule> rules = getDefaultRules();
+        
+        if (tierInstance.getTier().getProductReleases() != null) {
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: Please optimise the code on this line, extract the content of tierInstance.getTier().getProductReleases() in a auxiliary variable.

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 33]
```diff
@@ -522,5 +516,180 @@ public void update(ClaudiaData claudiaData, TierInstance tierInstance, Environme
...
+            return new Rule(protocol, port, port, "", "0.0.0.0/0");
+        }
+
+    }
+    
+    private void removeSecurityGroup (ClaudiaData data, TierInstance tierInstance) 
+    		throws InvalidEntityException, InfrastructureException {
+    	log.info("Deleting security group " + tierInstance.getSecurityGroup().getName() 
+    			+ " in tierInstance " + tierInstance.getName());
+        if (tierInstance.getSecurityGroup() != null && !tierInstance.getVdc().isEmpty()) {
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: the code is not well indented.

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java [hunk 34]
```diff
@@ -522,5 +516,180 @@ public void update(ClaudiaData claudiaData, TierInstance tierInstance, Environme
...
+    }
+    
+    /**
+     * @param infrastructureManager
+     *            the infrastructureManager to set <property name="tierInstanceDao" ref="tierInstanceDao"/>
+     */
+    public void setInfrastructureManager(InfrastructureManager infrastructureManager) {
+        this.infrastructureManager = infrastructureManager;
+    }
+
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: Add comments (javadoc) to the setters

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/util/OpenStackSyncImpl.java [hunk 35]
```diff
@@ -131,13 +131,13 @@ public void syncronize(Connection connection, boolean onecycle) throws OpenStack
...
                     try {
                         // Synchronizing secGroups
-
-                        securityGroupsSync(user.getToken(), user.getTenantId());
+                        //securityGroupsSync(user.getToken(), user.getTenantId());
                         // Synchronizing instances
                         tierInstancesSync(claudiaData);
-                    } catch (OpenStackException e) {
+                    } catch (OpenStackSynchronizationException e) {
                         String message = "Error accesing OpenSatck. Token outof date for user " + user.getUserName();
```
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/util/OpenStackSyncImpl.java]: Error accessing OpenSatck. Token out of date for user

File: model/src/main/java/com/telefonica/euro_iaas/paasmanager/model/TierInstance.java [hunk 36]
```diff
@@ -85,6 +85,9 @@
     @JoinTable(name = "tierinstance_has_networkinstance", joinColumns = { @JoinColumn(name = "tierinstance_ID", nullable = false, updatable = false) }, inverseJoinColumns = { @JoinColumn(name = "networkinstance_ID", nullable = false, updatable = false) })
```
Comment (flopezag) [inline model/src/main/java/com/telefonica/euro_iaas/paasmanager/model/TierInstance.java]: gt 120 chars.

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java (hunk 34 above)
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: NTC

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java (hunk 21 above)
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: NTC

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java (hunk 33 above)
Comment (flopezag) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/TierInstanceManagerImpl.java]: NTC

**Comment by flopezag** [comment]
LGTM

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/TierDao.java [hunk 37]
```diff
@@ -95,4 +95,13 @@ Tier loadTierWithProductReleaseAndMetadata(String tierName, String vdc, String e
...
     Tier loadComplete(Tier newTier) throws EntityNotFoundException;
+    
+    /**
+     * Update the securitygroupId of a Tier
+     * @param tier
+     * @param idSecurityGroup
+     * @return
+     * @throws InvalidEntityException
+     */
+    //Tier updateTierSecurityGroupId(Tier tier, String idSecurityGroup) throws InvalidEntityException;
```
Comment (hmunfru) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/dao/TierDao.java]: if it is commented, please remove the code

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java [hunk 38]
```diff
@@ -419,8 +428,29 @@ private void deletePaaSDB(ClaudiaData claudiaData, EnvironmentInstance envInstan
         } catch (Exception e) {
             log.error(e.getMessage());
             error = true;
+        } finally {
+            //Deleting SG
+            log.info("Deleting security group from tierInstance " + tierInstance.getName() + " in TierInstance");
+            SecurityGroup secGroup = tierInstance.getSecurityGroup();
+            if (secGroup != null && !tierInstance.getVdc().isEmpty()) {
```
Comment (hmunfru) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/EnvironmentInstanceManagerImpl.java]: por que el vdc puede ser vacio? vdc  = tenant

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/SecurityGroupManagerImpl.java [hunk 39]
```diff
@@ -48,25 +48,78 @@
...
-
-        if (securityGroup.getRules() != null) {
-            for (Rule rule : securityGroup.getRules()) {
-                rule.setIdParent(idSecurityGroup);
-                rule = ruleManager.create(region, token, vdc, rule);
-                securityGroupDB.addRule(rule);
-            }
+        
+        List<Rule> rules = securityGroup.getRules();
+        //if (securityGroup.getRules() != null) {
```
Comment (hmunfru) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/SecurityGroupManagerImpl.java]: nunca va a ser nulo?

File: core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/SecurityGroupManagerImpl.java [hunk 40]
```diff
@@ -48,25 +48,78 @@
...
+     * @param region
+     * @param token
+     * @param vdc
+     * @param securityGroup
+     * @return
+     * @throws AlreadyExistsEntityException
+     * @throws InvalidEntityException
+     * @throws InfrastructureException
+     */
+    public SecurityGroup createInOpenstack(String region, String token, String vdc, SecurityGroup securityGroup)
```
Comment (hmunfru) [inline core/src/main/java/com/telefonica/euro_iaas/paasmanager/manager/impl/SecurityGroupManagerImpl.java]: para que existe este metodo? ya lo tienes implementado en el metood anterior

**Comment by jframos** [comment]
LGTM. To be tested on Experimentation Environment.

**Pull request merged**